Tools 250 curation qa ntbok 53

jychien opened the pull request:
Updated curation notebooks for the following:

- establish reasonable thresholds for gene counts for each organism & include those checks
- evaluate_sparsity() - looks like the check is correct (anything 50% sparse or more has to be sparse csr_matrix) but should be ERROR now instead of WARNING
- check that the gene IDs are from the organism in obs - ERROR for any with mixed organism in obs
- curation_visium.ipynb:
  - addition to adata.obs['assay_ontology_term_id'] to list out other possible assays for V2
  - update if adata.obs.shape[0] < 4992 to be assay specific

Current logic:
- if there is more than a single organism according to either the indices in var or organism_ontology_term_id in obs, `evaluate_var_df()` will report the error and then not go through with with number of genes. Please let me know if this behavior should be updated.
- the assumption is that adata.var and adata.raw.var will be matching, so looks only at adata.var

Things checked:
- Ran notebook using a correct AnnData, resulted in `GOOD:Single organism found: {'NCBITaxon:9606'}` and `GOOD:36387 genes out of possible 39583 present`
- Ran notebook using AnnData with < threshold genes for human, resulted in `ERROR:only 14000 out of 39583 10x biotype genes, results in 0.353687188944749 (0.40 threshold)`
- Ran notebook using AnnData with < threshold genes for human, resulted in `WARNING:17000 out of 39583 10x biotype genes, results in 0.42947730086148095 (0.60 threshold)`
- Ran notebook using AnnData with CSC matrix, resulted in `ERROR:X should be converted to csr sparse`
- Used AnnData with mixed organisms (mouse and human), resulted in: `ERROR:Multiple organisms found in obs metadata: ['NCBITaxon:9606', 'NCBITaxon:10090']`

jahilton commented on cellxgene_resources/cellxgene_mods.py:
```diff
@@ -888,3 +897,61 @@ def evaluate_donors_sex(adata):
+    elif fraction < 0.6:
+        report(f'{len(adata.var.index)} out of {num_genes_biotype} 10x biotype genes, results in {fraction} (0.60 threshold)','WARNING')
+    else:
+        report(f'{len(adata.var.index)} genes out of possible {num_genes_biotype} present', 'GOOD')
```
I don't know if "possible" is the correct term here (Would be odd to see "60k genes out of possible 33k present")
Maybe "x genes present, compared against y genes"?
Maybe also display the fraction as that's what is being weighed?

jychien replied:
How about this:
`report(f'{len(adata.var.index)} genes present, compared against {num_genes_biotype} 10x biotype genes; fraction: {fraction}', 'GOOD')`

jahilton commented on cellxgene_resources/cellxgene_mods.py:
```diff
@@ -888,3 +897,61 @@ def evaluate_donors_sex(adata):
+
+
+    # Check that this is single organism both in metadata and var index, exit function as does not make sense to check genes with multiple organisms
+    obs_organisms = adata.obs['organism_ontology_term_id'].unique()
```
cleaner reporting if this is `.tolist()`

jahilton commented on cellxgene_resources/cellxgene_mods.py:
```diff
@@ -888,3 +897,61 @@ def evaluate_donors_sex(adata):
+        report(f'Multiple organisms found in obs metadata: {obs_organisms}', 'ERROR')
+        return
+    elif len(var_organisms) > 1:
+        report(f'Multiple organisms found in var index: {obs_organisms}', 'ERROR')
```
prints obs_organisms but should be var_organisms

also `NCBITaxon:2697049` (covid) should be OK as an additional organism in var

jychien replied:
Thanks for catching this.

jychien commented on cellxgene_resources/cellxgene_mods.py:
```diff
@@ -196,22 +198,29 @@ def determine_sparsity(x):
+            valid = False
+
+    if valid:
+        report('all matrices are csr', 'GOOD')
```
Yes! Oops, I'll update the wording on the report. I'll have it just print an "all clear", but let me know if you would like it to evaluate what types of matrices are in the h5ad.

brian-mott commented on cellxgene_resources/cellxgene_mods.py:
```diff
@@ -10,9 +10,11 @@
 import squidpy as sq
 import subprocess
 import sys
+import requests
```
is requests used anywhere? looks like it was inadvertently added

brian-mott commented on cellxgene_resources/cellxgene_mods.py:
```diff
@@ -888,3 +897,63 @@ def evaluate_donors_sex(adata):
+
+    # Check that this is single organism both in metadata and var index, exit function as does not make sense to check genes with multiple organisms
+    obs_organisms = adata.obs['organism_ontology_term_id'].unique().tolist()
+    var_organisms = {gencode.get_organism_from_feature_id(id).value for id in adata.var.index.to_list()}
```
`get_organism_from_feature_id()` returns a `SupportedOrganism` enum or `None`, so if `evaluate_var_df()` is called when `adata.var` is indexed on gene symbol and/or deprecated Ensemble IDs have not been filtered, then this will raise an `AttributeError`

There could be logic to deal with this, try to catch this exception and print an error to index on ID and filter, or place the call to `evaluate_var_df` further in the curation notebook when var is expected to be indexed and filtered appropriately

jychien replied:
Good point. For human, there are some genes that have Ensembl ID for both gene name and gene ID. So, I will check that None does not exist in var_organisms. Let me know if there is some edge case that I am not catching.

jychien replied:
Sounds good! Will return early and exit out of evaluate_var_df() contains gene symbols or any Ensembl IDs that are not on the approved gene lists.

brian-mott commented on cellxgene_resources/cellxgene_mods.py:
```diff
@@ -888,3 +897,63 @@ def evaluate_donors_sex(adata):
+        if 'NCBITaxon:2697049' not in obs_organisms:
+            report(f'Multiple organisms found in obs metadata: {obs_organisms}', 'ERROR')
+            return
+    elif len(var_organisms) > 1:
```
Might also be nice to rearrange this section to report both obs and var organism errors then return early. When there's multiple organisms in obs and var, then only the error report for obs will show; `evaluate_var_df` will have to be called again with obs fixed to get the var organism error

jychien replied:
Good point. I've updated logic, and will check both obs and var before breaking out of evaluate_var_df if there is an error.

jychien commented:
@jahilton @brian-mott @Jchaffer787 Looks like I've missed many edge cases. Thank you for all the comments. I'll take a deep dive today and some out with something more robust tomorrow. If you find any more issues, please add to the comments.

Jchaffer787 commented:
Nothing to add regarding code, looks like I was a little late to the party.
I was able to run the private fly dataset through as a test and output looks good:
```
GOOD:Single organism found: {'NCBITaxon:7227'}
GOOD:12553 genes present, compared against 13986 10x biotype genes; fraction: 0.8975403975403975
```
Could maybe be a space after 'GOOD:', but that's certainly a nitpick.

Jchaffer787 reviewed (commented):
Nothing to add regarding code, looks like I was a little late to the party.
I was able to run the private fly dataset through as a test and output looks good:
```
GOOD:Single organism found: {'NCBITaxon:7227'}
GOOD:12553 genes present, compared against 13986 10x biotype genes; fraction: 0.8975403975403975
```
Could maybe be a space after 'GOOD:', but that's certainly a nitpick.

jychien commented on cellxgene_resources/cellxgene_mods.py:
```diff
@@ -888,3 +897,63 @@ def evaluate_donors_sex(adata):
+            report(f'Multiple organisms found in var index: {var_organisms}', 'ERROR')
+            return
+    else:
+        report(f'Single organism found: {var_organisms}', 'GOOD')
```
Will need to incorporate logic for Covid. If var is covid+human and obs is just human, then still need to catch this error.

jychien commented:
Alrighty. Here are things updated:

- Dropped import requests. Thanks for catching, I initially was going to go through gencode files in the single cell repo.
- Fixed Visium_assays embeddings and image check
- Added space after "GOOD:|ERROR:|WARNING:"
- Multiple organisms not results on check for both var and obs:
> ERROR: Multiple organisms found in obs metadata: ['NCBITaxon:9606', 'NCBITaxon:10090']
ERROR: Multiple organisms found in var index: ['NCBITaxon:10090', 'NCBITaxon:9606']

- If non approved gene in var index, evaluate_var_df() will report and skip any other checks:
> ERROR: Features in var.index are gene symbols and/or contain deprecated Ensembl IDs

- Allow for covid genes in var and still have it be considered single organism:
> There are covid genes present in var
GOOD: Single organism found: ['NCBITaxon:9606']

- Will ERROR is organism between var and obs are different:
> ERROR: Different organisms found between var index and obs metadata: NCBITaxon:9606, NCBITaxon:10090

- Will ERROR Is there is covid in obs metadata:
> ERROR: Covid is found in obs metadata

jahilton reviewed (approved):
lgtm

brian-mott reviewed (approved):
Looks good to me too, thanks Jenny for all the hard work!

Jchaffer787 commented:
LGTM! Good output from fly file:
```
GOOD: Single organism found: ['NCBITaxon:10090']
GOOD: 22354 genes present, compared against 33989 10x biotype genes; fraction: 0.6576833681485187
```
Great work Jenny!!
